sky/shell/platform/android/org/domokit/sky/shell/PlatformViewAndroid.java

```diff
@@ -479,4 +495,63 @@ private boolean handleAccessibilityHoverEvent(MotionEvent event) {
         return true;
     }
 
+    /**
+     * Send a message to the Flutter application.  The Flutter Dart code can register a
+     * host message handler that will receive these messages.
+     */
+    public void sendToFlutter(String message, final MessageReplyCallback callback) {
+        mEmbeddedApplication.sendToFlutter(message,
+            new EmbeddedApplication.SendToFlutterResponse() {
+                @Override
+                public void call(String reply) {
+                    if (callback != null) {
+                        callback.onReply(reply);
+                    }
+                }
+            });
+    }
+
+    public void sendToFlutter(String message) {
+        sendToFlutter(message, null);
+    }
+
+    /** Callback invoked when the app replies to a message sent with sendToFlutter. */
+    public interface MessageReplyCallback {
+        void onReply(String reply);
+    }
+
+    /**
+     * Register a callback to be invoked when the Flutter application sends a message
+     * to its host.
+     */
+    public void setOnMessageListener(OnMessageListener listener) {
+        mOnMessageListener = listener;
+    }
+
+    public interface OnMessageListener {
+        /**
+         * Called when a message is received.
+         * @return the reply to the message (can be null)
+         */
+        String onMessage(String message);
```

Comment:
Actually I think all that's required is to pass the SendToHostResponse down to the message listener. The API just above this one is already structured asynchronously.